### ISSUE TITLE
'open containing folder' in TOPPAS not working under OS X

### ISSUE THREAD

**Issue by lars20070** [OpenMS/OpenMS#1299]
By right-clicking an output node, TOPPAS allows to open the folder containing the result file. This option works fine under Windows, but not under OS X 10.10.2.

**Comment by timosachsenberg** [comment]
Which OpenMS?

**Comment by lars20070** [comment]
Current develop.

**Comment by cbielow** [comment]
we've had this before:
https://github.com/OpenMS/OpenMS/commit/b48c4a875f190bad126cd346b8b0e48dcb0b0e0b

BTW: the commit from above introduces memory leaks by allocating QProcess but never free'ing it....
(A normal QProcess on stack would do -- no need for new())

**Comment by cbielow** [comment]
@lars20070  Do you see the the error message as written in the above commit?!

**Comment by lars20070** [comment]
@cbielow No error message in the log window.

**Comment by cbielow** [comment]
So it just does "nothing"?

according to
http://doc.qt.digia.com/qq/qq18-macfeatures.html
```
 QProcess process;
 process.start("/usr/bin/open", QStringList() << path, QIODevice::ReadOnly);
 process.waitForFinished();
```
should work.
More or less what we are doing.
ReadMode is different, and maybe we should do waitForFinished() as well. I'm not sure if 'open' blocks or returns.
Someone with a MacOSX has to try :)

**Comment by cbielow** [comment]
Also, getting rid of the Code duplication would be nice.

**Comment by lars20070** [comment]
Slightly off topic: Double clicking *.toppas opens a workflow in TOPPAS. Works fine under Windows. In OS X an additional blank tap opens.

@cbielow I will try your code changes later this week.

**Comment by lars20070** [comment]
@cbielow The new code compiles but does not change the behaviour.

https://github.com/lars20070/OpenMS/blob/fix/TOPPAS/src/openms_gui/source/VISUAL/TOPPASInputFileListVertex.cpp#L227

**Comment by cbielow** [comment]
try
waitForFinished()


If that does not help, try the DesktopServices (i.e. delete the MacOSX 
specific part)-- we'll have to figure out then why the fix was 
introduced in the first place, but lets worry about that later.

And then there is:
http://stackoverflow.com/questions/3490336/how-to-reveal-in-finder-or-show-in-explorer-with-qt


Am 3/11/2015 um 2:11 PM schrieb Lars Nilse:

**Comment by timosachsenberg** [comment]
Any news?
By the way - I added this manual test to our how to....
https://github.com/OpenMS/OpenMS/wiki/Testing-an-OpenMS-release-candidate

**Comment by lars20070** [comment]
@timosachsenberg Removing Stephan's original changes  https://github.com/OpenMS/OpenMS/commit/b48c4a875f190bad126cd346b8b0e48dcb0b0e0b does the trick. Solved in #1307.

**Comment by jpfeuffer** [comment]
In my case. It only doesnt work on the "output file" node. If I try it on the preceding node (e.g. FileInfo) it correctly opens the temporary folder. I dont know if you still need that information.

**Comment by cbielow** [comment]
@jpfeuffer Which OS?

On my box (Win7) all nodes work.
Did your pipeline finish? If not, the output folder might not even be present yet.

**Comment by timosachsenberg** [comment]
Mac is the problem.

**Comment by jpfeuffer** [comment]
@cbielow It was MacOS 10.10.2. And yes the pipeline was finished (folder not empty). I rechecked.

**Comment by lars20070** [comment]
Resolved in #1307.

**Comment by timosachsenberg** [comment]
please let's wait for a bit more testing when we have nightly installers that contain the fix

**Comment by timosachsenberg** [comment]
Marc and Johannes did test the installer and the problem doesn't seem to be resolved:

> Damn it. On 10.8.5, I now get this message box "The folder XY could not be opened.".
> 
> But it *did* work when I compiled Lars' branch, so maybe it could be a problem with the permissions of the TOPPAS.app? Some setting in some Info.plist or something?
> 
> To confirm this, I just tried to run TOPPAS not via the MacOS mechanism by clicking TOPPAS.app but from the command line. The result is promising: It *does* work when I just run /Applications/OpenMS-2.0.0/TOPPAS.app/Contents/MacOS/TOPPAS from the command line…
> 
> Cheers,
> Hannes

@aiche does this ring a bell? was that the original reason to introduce the mac specific code?

**Comment by aiche** [comment]
@timosachsenberg Sorry Timo, can't remember the exact error message etc. the fact that it does work from manual build/executable but not from the package somehow sounds familiar. We had quite a lot of problems with the installer packages on mac os x. Could you verify if the old version works with installer?

**Comment by hannesveit** [comment]
@aiche You mean whether TOPPAS from the 1.11.1 MacOS installer can open the containing folder when started via launchd? Yes, that works.

To me it looks as if launchd under 10.8 and older somehow forbids TOPPAS to launch another process (Finder)  via the Qt mechanism to open that directory. For some reason, this is not the case anymore in newer versions of MacOS while newer versions don't support the old way. I've been browsing the Apple docs for an appropriate parameter to set in the Info.plist but didn't really find anything suitable…

Could we maybe just check the MacOSX version at runtime and then choose one way or the other?

**Comment by hannesveit** [comment]
MacVersion QSysInfo::macVersion() would do the job.

Objections?

**Comment by timosachsenberg** [comment]
nope but we should check it on 10.9 first

**Comment by aiche** [comment]
@hannesveit `OpenMSInfo.cpp` implements the extraction of the OS X version based on the `Gestalt` API of OS X.  In the line of @cbielow 

> Also, getting rid of the Code duplication would be nice.

We could bundle the functionality to open a folder in a single function (e.g., `File::openFolder`). There one could also implement the switch based on the OS X version. 

The only downside is that `Gestalt` is deprecated and yes `QSysInfo::macVersion()` but gives some headaches since we have different Qt versions with differing enum values (e.g., Qt 4.8.6 includes `QSysInfo::MV_10_10` but Qt 4.7 only has values up to `QSysInfo::MV_10_5`)

**Comment by hannesveit** [comment]
Ouch.

**Comment by aiche** [comment]
digging a bit deeper in the Qt code, they actually use Objective-C code to determine the exact system version of OS X and then reduce it to the given enum. Not sure wether it is good idea, but you could try to "port" the qt objc function (`const QAppleOperatingSystemVersion version = qt_apple_os_version();`) to OpenMS.

**Comment by timosachsenberg** [comment]
I just checked on our 10.9 VM:
Old installer 1.11.1 works
New installer (nightly from master) fails
Maybe we should just create two installer like we did before - one for pre 10.10 and one for 10.10+?

**Comment by aiche** [comment]
But then we would have to keep both code paths and enable based on a configure switch

**Comment by timosachsenberg** [comment]
Honestly, I think the issue is not big enough to stale the release and I would prefer to just create two installers. Don't get me wrong - we should definitly keep the issue open and hope to find a solution soon but the impact of threating this as a "known issue" is very limited. 
Only ones users that compile their own OpenMS and package it to install it on a computer running 10.8 and 10.9 will have the problem if I understood all implications correctly.
For now I would just keep a "legacy" branch parallel to master that contains the old code and create binaries from it.
I know it is not ideal but considering more pressing issues or bugs I think this is totally valid to do so.

**Comment by aiche** [comment]
+1 for the solution-oriented approach ;)

**Comment by hannesveit** [comment]
+1

**Comment by cbielow** [comment]
Are the enums across Qt versions consistent in their value when casted to Int? So even if the value is not known for a particular version, we can use Qt to check the macosx version?
-- 
Sent via mobile phone.

Am 27. März 2015 16:23:30 MEZ, schrieb Johannes Veit <notifications@github.com>:

**Comment by cbielow** [comment]
how about
```
sw_vers -productVersion 
```
Seems to be present in all relevant versions. not sure about the '-productVersion' option though.

**Comment by hannesveit** [comment]
It seems like Timo spotted and fixed the bug!

There's a new installer for Mac: http://ftp.mi.fu-berlin.de/pub/OpenMS_ReleaseCandidates/MacLegacy_10.8_9/OpenMS-2.0.0-master_possfix.dmg

which fixes the problem for me on OSX 10.8.5. Since I used to experience the same symptoms Julianus described above for 10.10 (works on input- and tool nodes but not on output nodes), chances are pretty high that with this fix we will need only one installer for Mac.

Can a 10.10 user confirm?

**Comment by jpfeuffer** [comment]
Yes, I can confirm that the bug is solved for OS X 10.10 with this installer.
